File: LeetCode/iter1/python/LowestCommonAncestorOfABinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {TreeNode} p
    # @param {TreeNode} q
    # @return {TreeNode}
    def lowestCommonAncestor(self, root, p, q):
        if(root == None):
            return None
        
        pVal = p.val
        qVal = q.val
        while(True):
            rootVal = root.val
            if(pVal<rootVal and qVal<rootVal): #都在根节点的左边，访问根的左子树
                root = root.left
            elif(pVal>rootVal and qVal>rootVal):  #都在根节点的右边，访问根的右子树
                root = root.right
            else: #在两边则返回结果
                return root

File: LeetCode/iter2/C++/LowestCommonAncestorOfABinarySearchTree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <queue>
#include <unordered_map>
#include <unordered_set>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:

	TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
		if (root == NULL)
			return NULL;

		if (p == NULL && q == NULL)
			return root;
		else if (p == NULL && q != NULL)
			return q;
		else if (p != NULL && q == NULL)
			return p;
		else if (p->val == root->val)
			return p;
		else if (q->val == root->val)
			return q;
		else if (p->val < root->val && q->val < root->val)
			return lowestCommonAncestor(root->left,p,q);
		else if (p->val > root->val && q->val > root->val)
			return lowestCommonAncestor(root->right, p, q);
		else
			return root;
	}
};

int main()
{
	return 0;
}